Shu11698/vector tiled layer

shubham7109 opened the pull request:
Vector Tiled. Layer Sample

- Sets map with a given ArcGISVectorTiledLayer
- Currently the ExportVectorTilesJob returns a Job.Status.FAILED [[Link](https://github.com/Esri/arcgis-runtime-samples-android/blob/70ac98a2a00372482e80f772fc24e917bcd77e08/kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt#L149)]. Due to the code expects the Item resource cache directory path to be an empty directory.
- ~~Settings the Basemap sometimes would throw: Object is already owned. [[Link](https://github.com/Esri/arcgis-runtime-samples-android/compare/shu11698/vector-tiled-layer?expand=1#diff-dc57b5914a3aed30c1d6dfe34274a94b31b0cfab3e2001e703dee38bc68abe2cR140)]~~

shubham7109 commented:
Updates: 
- Fixed tile packages issue. Code now functions without bugs/crashes

Todo:
- Clear style resource files on app launch
- Download `.vtpk` file programmatically
- Fix the Basemap `layer.copy()` workaround.

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+        itemResourceCache.addDoneLoadingListener{
+            if(itemResourceCache.loadStatus == LoadStatus.LOADED){
+                setResourceAndVectorTileCache(itemResourceCache)
+            }
+            else{
```
```suggestion
            } else {
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+
+    }
+
+    private fun setResourceAndVectorTileCache(itemResourceCache: ItemResourceCache){
```
A function level comment here would be helpful

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+import com.esri.arcgisruntime.sample.vectortiledlayercustomstyle.databinding.ActivityMainBinding
+import kotlinx.android.synthetic.main.activity_main.*
+
+//[DocRef: END]
```
This probably snuck in from another sample and should be deleted

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/res/layout/activity_main.xml:
```diff
@@ -0,0 +1,45 @@
+    android:layout_height="match_parent"
+    tools:context=".MainActivity">
+
+
```
can remove one line of whitespace here

TADraeseke replied:
still whitespace here

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/res/layout/spinner_item.xml:
```diff
@@ -0,0 +1,34 @@
+    android:layout_height="wrap_content"
+    android:padding="10dp"
+    android:orientation="horizontal">
+
```
delete whitespace

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,4 @@
+<resources>
+    <string name="app_name">vector-tiled-layer</string>
```
```suggestion
    <string name="app_name">ArcGIS vector tiled layer (custom style)</string>
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/res/layout/spinner_item.xml:
```diff
@@ -0,0 +1,34 @@
+
+
+    <View
+        android:id="@+id/color_view"
```
ids are usually camelCase, and layout file names snake_case in Android...don't ask me why 🤣

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+        super.onDestroy()
+    }
+
+    //[DocRef: END]
```
can delete this line

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+    class CustomDropDownAdapter(private val context: Context) : BaseAdapter() {
+
+        // The names of the each Vector Tiled Layer
+        private val styleNames = arrayOf("Default", "Style 1", "Style 2", "Style 3", "Offline custom style: Light", "Offline custom style: Dark")
+
+        // The drawable XML file names for the associated style name
+        private val styleDrawableNames = arrayOf("default_color","style1_color","style2_color","style3_color","custom1_color","custom2_color")
```
The linter might ask you to put these in the strings.xml file, which probably isn't a bad idea in this case

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+
+        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
+            val view: View
+            val vh: ItemHolder
```
Maybe a more readable variable name viewHolder or itemHolder or something

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+
+    //[DocRef: END]
+
+    class CustomDropDownAdapter(private val context: Context) : BaseAdapter() {
```
A class level comment might clear things up for our users here

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+        }
+    }
+
+    //[DocRef: Name=Pause and resume-Android, Category=Get started, Topic=Develop your first map app with Kotlin]
```
can delete this line

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+
+                // Sets the selected itemID to either the Online/Custom ID.
+                currentItemID = when (position) {
+                    4 -> {
```
I understand what you're doing here, but a comment to explain why you're only checking for positions 4 and 5 would be helpful here

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+
+        // Set the currentItemID to default layer.
+        currentItemID = onlineItemIds[0]
+    }
```
create a map online once in onCreate 

`        mapView.map = ArcGISMap()`

and remove it from the setMap(...) function

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,293 @@
+     * Set the map using the layer and the viewpoint.
+     */
+    private fun setMap(layer: ArcGISVectorTiledLayer, viewpoint: Viewpoint){
+        //Reset the map to release resources
+        mapView.map = null
+        //mapView.map.basemap = null
+
+        // Assign a new map created from the base layer.
+        val basemap = Basemap(layer.copy())
+        val map = ArcGISMap(basemap)
+        mapView.map = map
```
```suggestion
        mapView.map.basemap.baseLayers.clear()
        mapView.map.basemap.baseLayers.add(layer)
```

Something like this gets around the `Object Owned` error. We don't want to create a new map or basemap every time, I think we just want to replace the layer on the existing map's basemap

TADraeseke reviewed (changes requested):
Thanks Shubham. Awesome work getting through this tough first sample!

A couple things to do to make your life easier on these PRs.

1) In Android Studio run Code -> Reformat code on the MainActivity and xml files
2) Run Analyze -> Inspect code to check for linting errors

Within the sample, we'll also want to standardize the comments to either `// Use sentences.` or `// just be notes`. Don't mind which you prefer, but let's just make it standard within the sample

Nice work with this again! Why don't you implement the fixes you're happy to make on your own and we can chat on a call through any that you have questions or issues with.

Thanks! 👍

shubham7109 commented:
Updates:

- Resolved comments from @TADraeseke
- Created `README.md` and `README.metadata.json`
- Need to update the `provision_from`, `provision_to`, and `redirect_from` in the metadata file.

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.md:
```diff
@@ -0,0 +1,44 @@
+
+Custom styled ArcGIS vector tiled layer             |  ArcGIS vector tiled layer list             |  Offline custom style
+:-------------------------:|:-------------------------:|:-------------------------:
+![Custom styled ArcGIS vector tiled layer](vector-tiled-layer-style.png)  |  ![ArcGIS vector tiled layer list](vector-tiled-layers-list.png)  |  ![Offline custom style](vector-tiled-layer-custom-2.png)
```
I know the common design has two screenshots, but we don't support more than one one in the android sample viewer. Just use the first. It also needs to match the title like:

![Image of ArcGIS vector tiled layer custom style]arcgis-vector-tiled-layer-custom-style.png

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.md:
```diff
@@ -0,0 +1,44 @@
+    iv. Start the job using  `ExportVectorTilesJob.start()`.  
+    v. Construct an `VectorTileCache` using the name of the local vector tile package.  
+    vi. Once the job is complete, construct an `ArcGISVectorTiledLayer` using the vector tile cache and the `ItemResourceCache` from the job's result.  
+3. Create an `Basemap` from the `ArcGISVectorTiledLayer`.
```
```suggestion
3. Create a `Basemap` from the `ArcGISVectorTiledLayer`.
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Relevant API
+
+* ArcGISVectorTiledLayer
+* ExportVectorTilesTask
+* ItemResourceCache
+* Map
+* VectorTileCache
```
```suggestion
* ArcGISVectorTiledLayer
* ExportVectorTilesTask
* ItemResourceCache
* VectorTileCache
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Offline data
+
+This sample uses the [Dodge City OSM](https://www.arcgis.com/home/item.html?id=f4b742a57af344988b02227e2824ca5f) vector tile package. It is downloaded from ArcGIS Online automatically.
```
```suggestion
## Offline data

1. Download the data from [ArcGIS Online](https://arcgisruntime.maps.arcgis.com/home/item.html?id=f4b742a57af344988b02227e2824ca5f).
1. Extract the contents of the downloaded zip file to disk.
1. Open your command prompt and navigate to the folder where you extracted the contents of the data from step 1.
1. Execute the following command:

`adb push dodge_city.vtpk /Android/data/com.esri.arcgisruntime.sample.arcgisvectortiledlayercustomstyle/files/dodge_city.vtpk`

Link | Local Location
---------|-------|
|[Dodge City VTPK](https://arcgisruntime.maps.arcgis.com/home/item.html?id=f4b742a57af344988b02227e2824ca5f)| `<sdcard>`/Android/data/com.esri.arcgisruntime.sample.arcgisvectortiledlayercustomstyle/files/dodge_city.vtpk |
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.metadata.json:
```diff
@@ -0,0 +1,46 @@
+{
+    "category": "TODO",
```
```suggestion
    "category": "Visualization",
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.metadata.json:
```diff
@@ -0,0 +1,46 @@
+    ],
+    "language": "kotlin",
+    "provision_from": [
+        "TODO"
```
```suggestion
        "https://arcgisruntime.maps.arcgis.com/home/item.html?id=f4b742a57af344988b02227e2824ca5f"
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.metadata.json:
```diff
@@ -0,0 +1,46 @@
+    ],
+    "provision_to": [
+        "TODO"
+    ],
+    "redirect_from": [
+        "TODO"
+    ],
+    "relevant_apis": [
```
```suggestion
    ],
    "relevant_apis": [
```

redirect not needed on new samples (this is used on the developer website to catch old samples that may have had their name changed)

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.metadata.json:
```diff
@@ -0,0 +1,46 @@
+        "ArcGISVectorTiledLayer",
+        "ExportVectorTilesTask",
+        "ItemResourceCache",
+        "Map",
```
```suggestion
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.metadata.json:
```diff
@@ -0,0 +1,46 @@
+        "VectorTileCache"
+    ],
+    "snippets": [
+        "src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt"
```
```suggestion
        "src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt"
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/build.gradle:
```diff
@@ -0,0 +1,32 @@
+    compileSdkVersion rootProject.ext.compileSdkVersion
+
+    defaultConfig {
+        applicationId "com.esri.arcgisruntime.sample.vectortiledlayercustomstyle"
```
```suggestion
        applicationId "com.esri.arcgisruntime.sample.arcgisvectortiledlayercustomstyle"
```

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/vectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,313 @@
+package com.esri.arcgisruntime.sample.vectortiledlayercustomstyle
```
```suggestion
package com.esri.arcgisruntime.sample.arcgisvectortiledlayercustomstyle
```

Sorry, one more name change across the sample

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/README.metadata.json:
```diff
@@ -0,0 +1,46 @@
+    "provision_from": [
+        "TODO"
+    ],
+    "provision_to": [
+        "TODO"
+    ],
```
```suggestion
    "provision_to": [
        "/dodge_city.vtpk"
    ],
```

TADraeseke reviewed (changes requested):
Thanks @shubham7109! Looking great!

Just a few small changes to the sample name required. I've included the details that should take care of getting the vtpk on the device for the sample. Probably best to discuss the details in call when you can.

Cheers!

shubham7109 commented:
Changes:
- Updated `README.metadata.json`
- Replaced package name to `arcgisvectortiledlayercustomstyle` in gradle, .kt and Manifest files
- Added `README.md` suggestions

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,27 @@
+        android:roundIcon="@mipmap/ic_launcher_round"
+        android:supportsRtl="true"
+        android:theme="@style/AppTheme">
+        <activity android:name=".MainActivity">
```
This still needs to be done

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,313 @@
+    }
+
+    /**
+     * Displays the layer of the given itemID.
```
Add param to function comments

https://github.com/Esri/arcgis-runtime-samples-android/blob/main/kotlin/trace-utility-network/src/main/java/com/esri/arcgisruntime/sample/traceutilitynetwork/MainActivity.kt#L219

shubham7109 replied:
Hey @TADraeseke. Thanks for this suggestion. intelliJ does not automatically create `@param` tags for Kotlin anymore since it is not recommended [Link](https://kotlinlang.org/docs/coding-conventions.html#avoid-redundant-constructs). Instead I will use `[]` tag to reference the arguments in the function description.

TADraeseke replied:
Oh wow great find! Let's try to move to this pattern going forward 👍

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,313 @@
+    /**
+     *  Checks local cache for local cache files.
+     *  If not it calls loadLayerWithOfflineCustomStyle() to retrieve the cache files.
+     */
```
param as above

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,313 @@
+
+    /**
+     * Retrieves the style resource files and caches it to the local device.
+     */
```
param as above

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,313 @@
+    /**
+     * Sets up an ArcGISVectorTiledLayer using the itemResourceCache
+     * for the setMap() function.
+     */
```
as above param

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,313 @@
+
+    /**
+     * Set the map using the layer and the viewpoint.
+     */
```
as above params

TADraeseke reviewed (changes requested):
Thanks @shubham7109 just a few small things now!


Need to update screenshot to 700 x 400 with no toolbar. png

Need to remove the dodge_city.vtpk file

shubham7109 commented:
Updates:

- Updated screenshot to 700 x 400 with no toolbar. png
- Removed the dodge_city.vtpk file
- Updated function level comments to use Kotlin [convention](https://kotlinlang.org/docs/coding-conventions.html#documentation-comments) instead of `param`

TADraeseke commented on kotlin/vector-tiled-layer-custom-style/src/main/java/com/esri/arcgisruntime/sample/arcgisvectortiledlayercustomstyle/MainActivity.kt:
```diff
@@ -0,0 +1,315 @@
+                val layer = ArcGISVectorTiledLayer(vectorTileCache, itemResourceCache)
+                layer.addDoneLoadingListener {
+                    if (layer.loadError != null)
+                        Log.e("VectorTiledLayer: ", layer.loadError.toString())
+                    else
+                        Log.d("VectorTiledLayer:", "Loaded successfully")
```
Let's pass the error as a toast as well and not bother with the loaded successfully debug line

TADraeseke reviewed (approved):
Just a few small changes and you're good to merge. Thanks!
